src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -0,0 +1,749 @@
+    assert(isSupertype(dep, mir.typ), s"not subtype: \n  $mir\n  ${ mir.typ }\n  $dep")
+    mir match {
+      case x@FilterColsIR(child, pred) =>
+        val (irDep, memo) = getDeps(pred, pred.typ, child.typ)
+        val child2 = rebuild(child, unionBaseType(child.typ, dep, irDep))
+        FilterColsIR(child2, rebuildIR(pred, child2.typ, memo))
+      case x@FilterRowsIR(child, pred) =>
+        val (irDep, memo) = getDeps(pred, pred.typ, child.typ)
+        val child2 = rebuild(child, unionBaseType(child.typ, dep, irDep))
+        FilterRowsIR(child2, rebuildIR(pred, child2.typ, memo))
+      case x@MatrixFilterEntries(child, pred) =>
+        val (irDep, memo) = getDeps(pred, pred.typ, child.typ)
+        val child2 = rebuild(child, unionBaseType(child.typ, dep, irDep))
+        MatrixFilterEntries(child2, rebuildIR(pred, child2.typ, memo))
+      case x@MatrixMapEntries(child, newEntries) =>
+        val (irDep, memo) = getDeps(newEntries, dep.entryType, child.typ)
+        val depMod = dep.copy(rvRowType = TStruct(dep.rvRowType.required, dep.rvRowType.fields.map { f =>
+          if (f.name == MatrixType.entriesIdentifier)
+            f.name -> f.typ.asInstanceOf[TArray].copy(elementType = irDep.entryType)
+          else
+            f.name -> f.typ
+        }: _*))
+        val child2 = rebuild(child, unionBaseType(child.typ, depMod, irDep))
+        MatrixMapEntries(child2, rebuildIR(newEntries, child2.typ, memo))
+      case x@MatrixMapRows(child, newRow, newKey) =>
+        val (irDep, memo) = getDeps(newRow, dep.rowType, child.typ)
+        val depMod = dep.copy(rvRowType = TStruct(irDep.rvRowType.fields.map { f =>
+          if (f.name == MatrixType.entriesIdentifier)
+            f.name -> unionType(child.typ.rvRowType.field(MatrixType.entriesIdentifier).typ,
+              f.typ,
+              dep.rvRowType.field(MatrixType.entriesIdentifier).typ)
+          else
+            f.name -> f.typ
+        }: _*), rowKey = child.typ.rowKey)
+        val child2 = rebuild(child, unionBaseType(child.typ, depMod, irDep))
+        MatrixMapRows(child2, rebuildIR(newRow, child2.typ, memo), newKey)
+      case x@MatrixMapCols(child, newCol, newKey) =>
+        val (irDep, memo) = getDeps(newCol, dep.colType, child.typ)
+        val child2 = rebuild(child, unionBaseType(child.typ, dep.copy(
+          colType = irDep.colType,
+          colKey = child.typ.colKey
+        ), irDep))
+        MatrixMapCols(child2, rebuildIR(newCol, child2.typ, memo), newKey)
+      case x@MatrixMapGlobals(child, newRow, value) =>
+        val (irDep, memo) = getDeps(newRow, dep.globalType, child.typ)
+        // fixme push down into value
+        val child2 = rebuild(child, unionBaseType(child.typ, dep.copy(globalType = irDep.globalType), irDep))
+        MatrixMapGlobals(child2, rebuildIR(newRow, child2.typ, memo, "value" -> value.t), value)
+      case x@MatrixRead(_, _, _, _, _) => upcast(x, dep)
```

Comment:
Don't we want to leave Read as-is and upcast the input to Write?  That seems to do the minimal amount of work.  (I realize there were some type difficulties that I aren't quite obvious from the code that I might not be remembering.)